UefiTestingPkg/AuditTests/PagingAudit/UEFI/Dxe/App/DxePagingAuditTestApp.c

```diff
@@ -405,113 +428,171 @@ PopulateEfiMemoryMap (
-    mEfiMemoryMap = (EFI_MEMORY_DESCRIPTOR *)AllocateZeroPool (mEfiMemoryMapSize);
+  SortMemoryMap (mEfiMemoryMap, mEfiMemoryMapSize, mEfiMemoryMapDescriptorSize);
+
+  return Status;
+}
+
+/**
+  Checks the allocation size of the global EFI memory map buffer and reallocates it to be 20%
+  larger if it is too small to hold the EFI memory map.
+
+  @retval EFI_SUCCESS   Global buffer is large enough or was reallocated.
+  @retval other         An error occurred while validating the global buffer.
+**/
+STATIC
+EFI_STATUS
+ValidateEfiMemoryMapSize (
+  VOID
+  )
+{
+  EFI_MEMORY_DESCRIPTOR  *EfiMemoryMap;
+  UINTN                  EfiMemoryMapSize;
+  UINTN                  EfiMemoryMapDescriptorSize;
+  UINTN                  EfiMapKey;
+  UINT32                 EfiDescriptorVersion;
+  EFI_STATUS             Status;
+
+  EfiMemoryMapSize           = 0;
+  EfiMemoryMap               = NULL;
+  EfiMemoryMapDescriptorSize = 0;
+  EfiMapKey                  = 0;
+  EfiDescriptorVersion       = 0;
+
+  Status = gBS->GetMemoryMap (
+                  &EfiMemoryMapSize,
+                  EfiMemoryMap,
+                  &EfiMapKey,
+                  &EfiMemoryMapDescriptorSize,
+                  &EfiDescriptorVersion
+                  );
+
+  if (Status != EFI_BUFFER_TOO_SMALL) {
+    UT_LOG_ERROR ("Failed to get the required EFI memory map size!\n");
+    return EFI_ABORTED;
+  }
+
+  if (EfiMemoryMapSize >= mEfiMemoryMapSize) {
+    FreeEfiMemoryMap ();
+    mEfiMemoryMapSize = EfiMemoryMapSize + (EfiMemoryMapSize / 5); // Increase size by 20%
+    mEfiMemoryMap     = AllocatePool (mEfiMemoryMapSize);
     if (mEfiMemoryMap == NULL) {
-      ASSERT (mEfiMemoryMap != NULL);
-      DEBUG ((DEBUG_ERROR, "%a - Unable to allocate memory for the EFI memory map.\n", __FUNCTION__));
+      UT_LOG_ERROR ("Failed to allocate %d bytes for EFI memory map!\n", mEfiMemoryMapSize);
       return EFI_OUT_OF_RESOURCES;
     }
+  }
 
-    Status = gBS->GetMemoryMap (
-                    &mEfiMemoryMapSize,
-                    mEfiMemoryMap,
-                    &EfiMapKey,
-                    &mEfiMemoryMapDescriptorSize,
-                    &EfiDescriptorVersion
-                    );
-    if (EFI_ERROR (Status)) {
-      FreePool (mEfiMemoryMap);
-    }
-  } while (Status == EFI_BUFFER_TOO_SMALL);
-
-  SortMemoryMap (mEfiMemoryMap, mEfiMemoryMapSize, mEfiMemoryMapDescriptorSize);
-
-  return Status;
+  return EFI_SUCCESS;
 }
 
 /**
-  Checks the input flat page/translation table for the input region and converts the associated
-  table entries to EFI access attributes.
+  Checks the input flat page/translation table for the input region and validates
+  the attributes match the input attributes.
 
   @param[in]  Map                 Pointer to the PAGE_MAP struct to be parsed
   @param[in]  Address             Start address of the region
   @param[in]  Length              Length of the region
   @param[out] Attributes          EFI Attributes of the region

```

Comment:
These are not the right parameters any longer

Reply:
Updated the input parmeters